### PR TITLE
Added clean exec functionality + ACI implementation - V2

### PR DESCRIPTION
**Pull request by erjadi**
Clean implementation of exec for Virtual Kubelet:

- Reusing standard Kubelet Executor interface (and remotecommand functions)
- Implemented ACI provider

Two TODOs are open:

- How to best pass namespace to provider (currently doing ACI specific way, should be changed)
- Expected streamcount is fixed number, but should be based on exec request to apiserver. However when using exec request, createStreams will wait indefinitely (for +1 stream).

This is a new pull request (invalidating https://github.com/virtual-kubelet/virtual-kubelet/pull/212) based on master after dependencies were fixed. 

People who are just interested in trying it out, use erjadi/virtual-kubelet:cleanexec from dockerhub.

### CODE REVIEW

File: providers/azure/aci.go [hunk 1]
```diff
@@ -329,6 +332,69 @@ func (p *ACIProvider) GetContainerLogs(namespace, podName, containerName string,
 	return logContent, err
 }
 
+// ExecInContainer executes a command in a container in the pod, copying data
+// between in/out/err and the container's stdin/stdout/stderr.
+func (p *ACIProvider) ExecInContainer(name string, uid types.UID, container string, cmd []string, in io.Reader, out, errstream io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize, timeout time.Duration) error {
+	cg, _, _ := p.aciClient.GetContainerGroup(p.resourceGroup, name)
```
Comment (robbiezhang) [inline providers/azure/aci.go]: check the error. If it failed, please close the out/err streams.

File: providers/azure/aci.go [hunk 2]
```diff
@@ -329,6 +332,69 @@ func (p *ACIProvider) GetContainerLogs(namespace, podName, containerName string,
...
+	terminalSize := remotecommand.TerminalSize{
+		Height: 60,
+		Width:  120,
+	}
+
+	if resize != nil {
+		terminalSize = <-resize // Receive terminal resize event if resize stream is present
+	}
+
+	xcrsp, _ := p.aciClient.LaunchExec(p.resourceGroup, cg.Name, container, cmd[0], terminalSize)
```
Comment (robbiezhang) [inline providers/azure/aci.go]: check error.

File: providers/azure/aci.go [hunk 3]
```diff
@@ -329,6 +332,69 @@ func (p *ACIProvider) GetContainerLogs(namespace, podName, containerName string,
...
+	// Cleanup on exit
+	defer c.Close()
+	defer out.Close()
+
+	if in != nil {
+		// Inputstream
+		go func() {
+			for {
+				var msg = make([]byte, 512)
+				_, err := in.Read(msg)
```
Comment (robbiezhang) [inline providers/azure/aci.go]: I tried the image, the exec works well, except one minor issue: it doesn't respond to "tab". Do you have any idea?

**Comment by robbiezhang** [comment]
@erjadi appreciate the great work. I tried your image, and it works well. I left some comments in the PR.
For the 1st TODO, I would recommend to add a GetPodFullName function in the interface to let the provider give the right format for the resource name. Can you elaborate the impact of the 2nd TODO?

File: providers/azure/aci.go (hunk 3 above)
Comment (erjadi) [reply providers/azure/aci.go]: Whoops you found a bug. I should check how many bytes were actually read from stdin. I was sending 512 bytes every time, mostly zeroes. Apparently that breaks tab functionality.

Fixed it, will push.

File: vkubelet/apiserver.go [hunk 4]
```diff
@@ -64,3 +68,36 @@ func ApiServerHandler(w http.ResponseWriter, req *http.Request) {
...
+		Stdin:  true,
+		Stdout: true,
+		Stderr: true,
+		TTY:    false,
+	}
+
+	idleTimeout := time.Second * 30
+	streamCreationTimeout := time.Second * 30
+
+	remotecommand.ServeExec(w, req, p, fmt.Sprintf("%s-%s", namespace, pod), "", container, []string{command}, streamOpts, idleTimeout, streamCreationTimeout, supportedStreamProtocols)
```
[the remark this answers is not in the document]
Comment (erjadi) [reply vkubelet/apiserver.go]: Sounds good, will implement it + provider stubs.

**Comment by erjadi** [comment]
Ok I fixed the following:
- Added error handling (closing the streams)
- Tab handling / proper buffer length
- Added GetPodFullName provider function + stubs

btw I noticed that not all the providers are instantiated in provider.go (huawei, azurebatch, mock, vic are missing). Should I add those in this PR or rather open a separate one?

**Comment by robbiezhang** [comment]
lgtm

**Pull request merged**